File: scripts/_dbg_cantata19_m11.ts
/**
 * Debug script: trace harmonic analysis for Cantata 19 Bach around m11 b4
 * Run: npx tsx scripts/_dbg_cantata19_m11.ts
 */
import fs from 'node:fs';
import {
  getRomanAnalysis,
  getActiveNotesTimeline,
  calculateNoteBeats,
  applyHarmonyRules,
  identifyChordCandidates,
  getKeySignature,
} from '../src/utils/musicTheory';

const raw = JSON.parse(fs.readFileSync('tests/Cantata 19 bach.json', 'utf8'));
const notes = raw.notes || raw.rawNotes || [];
const tonic = raw.keySignatureRoot || raw.keyTonic || 'Bb';
const isMinor = raw.isMinorMode || false;
const ts = raw.timeSignature || { numerator: 4, denominator: 4 };
const ctxs = raw.analysisContexts || [];

console.log('=== Cantata 19 Bach ===');
console.log('Tonic:', tonic, 'Minor:', isMinor, 'TS:', ts);

const keySig = getKeySignature(tonic, isMinor ? 'Minor' : 'Major');
applyHarmonyRules(notes, keySig, tonic, isMinor, ctxs, ts);
calculateNoteBeats(notes, ts);
const timeline = getActiveNotesTimeline(notes, ts);

// m10-m12: absBeat 36-47
for (const ev of timeline) {
  const ab = (ev as any).absBeat ?? (ev as any).absoluteBeat ?? 0;
  if (ab < 36 || ab > 47) continue;

  const allNotes = ev.notes || [];
  const structural = allNotes.filter(
    (n: any) => !n.isPassing && !n.isAppoggiatura && !n.isNeighbor
  );
  const pcs = [...new Set(structural.map((n: any) => n.midi % 12))].sort();
  const pitches = structural.map((n: any) => `${n.pitch || '?'}${n.octave || ''}`);

  // Analysis in Bb major (home key)
  const romanBb = getRomanAnalysis(structural, tonic, isMinor);
  // Analysis in Dm (suspected false tonicization)
  const romanDm = getRomanAnalysis(structural, 'D', true);
  // Analysis in F major (V region)
  const romanF = getRomanAnalysis(structural, 'F', false);

  // Chord candidates
  const candidates = identifyChordCandidates(structural);

  const mi = Math.floor(ab / 4) + 1;
  const beat = (ab % 4) + 1;

  console.log(
    `m${mi} b${beat} (abs=${ab.toFixed(1)})`,
    `pcs=[${pcs}]`,
    `notes=[${pitches}]`,
    `| Bb: ${romanBb?.roman || '?'}${romanBb?.figures || ''}`,
    `| Dm: ${romanDm?.roman || '?'}${romanDm?.figures || ''}`,
    `| F: ${romanF?.roman || '?'}${romanF?.figures || ''}`,
    `| cand: ${(candidates || []).map((c: any) => `${c.root}${c.type || c.quality || ''}`).join(', ') || 'none'}`
  );
}

File: scripts/_dbg_cantata19_m11.py
#!/usr/bin/env python3
"""Debug script: trace chords in Cantata 19 Bach to find Dm context source."""
import json, os

path = os.path.join(os.path.dirname(__file__), '..', 'tests', 'Cantata 19 bach.json')
with open(path) as f:
    d = json.load(f)

notes = d.get('notes', d.get('rawNotes', []))
print(f"File: Cantata 19 bach.json")
print(f"Tonic: {d.get('keySignatureRoot', d.get('keyTonic', '?'))}")
print(f"Minor: {d.get('isMinorMode', '?')}")
print(f"TS: {d.get('timeSignature', '?')}")
print(f"AnalysisContexts: {d.get('analysisContexts', [])}")
print(f"Total notes: {len(notes)}")
print()

NOTE_NAMES = ['C','Db','D','Eb','E','F','Gb','G','Ab','A','Bb','B']

def identify_chord(pcs_set):
    """Simple chord identification by pitch class set."""
    for root_pc in range(12):
        r = root_pc
        name = NOTE_NAMES[r]
        maj3 = (r + 4) % 12
        min3 = (r + 3) % 12
        p5 = (r + 7) % 12
        dim5 = (r + 6) % 12
        m7 = (r + 10) % 12
        
        if {r, maj3, p5}.issubset(pcs_set):
            if m7 in pcs_set:
                return f"{name}7"
            return f"{name}M"
        if {r, min3, p5}.issubset(pcs_set):
            if m7 in pcs_set:
                return f"{name}m7"
            return f"{name}m"
        if {r, min3, dim5}.issubset(pcs_set):
            return f"{name}dim"
    return f"?{sorted(pcs_set)}"

max_mi = max(n.get('measureIndex', 0) for n in notes)
print(f"Measures: {max_mi + 1}")
print()

# Show all measures
for mi in range(max_mi + 1):
    mn = [n for n in notes if n.get('measureIndex') == mi]
    if not mn:
        continue
    beats = sorted(set(n.get('beat') for n in mn))
    parts = []
    for b in beats:
        bnotes = [n for n in mn if n.get('beat') == b]
        pcs = set(nn.get('midi', 0) % 12 for nn in bnotes)
        chord = identify_chord(pcs)
        bass = min(nn.get('midi', 999) for nn in bnotes)
        bass_name = NOTE_NAMES[bass % 12]
        # Mark beats with A natural or C#
        flags = []
        if 9 in pcs:
            flags.append('A')
        if 1 in pcs:
            flags.append('C#')
        flag_str = f" [{','.join(flags)}]" if flags else ""
        parts.append(f"b{b}:{chord}/{bass_name}{flag_str}")
    
    marker = " <<<" if mi == 10 else ""  # m11 = measureIndex 10
    print(f"  m{mi+1}: {' | '.join(parts)}{marker}")

print()
print("=== Potential V/iii (A major = pcs 9,1,4) or V-vi in F (C->Dm) ===")
for mi in range(max_mi + 1):
    mn = [n for n in notes if n.get('measureIndex') == mi]
    if not mn:
        continue
    beats = sorted(set(n.get('beat') for n in mn))
    beat_chords = []
    for b in beats:
        bnotes = [n for n in mn if n.get('beat') == b]
        pcs = set(nn.get('midi', 0) % 12 for nn in bnotes)
        chord = identify_chord(pcs)
        beat_chords.append((b, chord, pcs))
    
    # Check consecutive beats for C->Dm pattern (V->vi in F = deceptive cadence)
    for i in range(len(beat_chords) - 1):
        b1, c1, pcs1 = beat_chords[i]
        b2, c2, pcs2 = beat_chords[i + 1]
        # C major/C7 followed by Dm
        if c1 in ('CM', 'C7') and c2 in ('Dm', 'Dm7'):
            print(f"  m{mi+1}: CONSECUTIVE C->Dm at b{b1}->b{b2} (V->vi in F, deceptive!)")
        # A major followed by Dm (V->I in Dm)
        if c1 in ('AM', 'A7') and c2 in ('Dm', 'Dm7'):
            print(f"  m{mi+1}: A->Dm at b{b1}->b{b2} (V->I in Dm!)")
